sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_azureappconfigurationprovider.py

```diff
@@ -416,64 +417,73 @@ def __init__(self, **kwargs) -> None:
             or self._keyvault_client_configs is not None
             or self._secret_resolver is not None
         )
+        self._current_refresh_check = False
         self._update_lock = Lock()
+        self._refresh_lock = Lock()
 
     def refresh(self, **kwargs) -> None:
         if not self._refresh_on:
             logging.debug("Refresh called but no refresh options set.")
             return
 
+        with self._refresh_lock:
```

Comment:
I think we might still block here if `refresh_lock` is acquired.

How about we use `refresh_lock.acquire(blocking=False)`.
This will return False if the lock is acquired. Then we don't need a `self._current_refresh_check`

Then in the try-except, we have a `finally` clause that releases the lock.

Reply:
This one's fine to be blocking as it is only two lines below so it's not at all a running operation. The with statement will auto unlock once it leaves, error or not.

The thing we need in the finally block is to unset the `_current_refresh_check` boolean, which we do.

Reply:
But then it seems the _current_refresh_check is now acting as our lock.